### PR TITLE
replace failure by thiserror

### PR DESCRIPTION
**Pull request by ariesdevil**
Use `thiserror` as the replacement of `failure` as `failure` is deprecated.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,23 +2,21 @@
...
+    feature = "cargo-clippy",
+    allow(verbose_bit_mask, unreadable_literal)
+)]
 #![cfg_attr(feature = "nightly", feature(test))]
 
-extern crate byteorder;
-extern crate bit_vec;
-#[macro_use]
-extern crate failure;
+use thiserror::Error as ThisError;
```
Comment (Flakebi) [inline src/lib.rs:11]: This should be a few lines down with the other uses.
I think you don’t even need to rename it, using it with `#[derive(Error, Debug)]` should work too.

File: src/lib.rs [hunk 2]
```diff
@@ -41,31 +39,26 @@ thread_local! {
         = RefCell::new(BitVec::from_elem(HASHTABLE_SIZE, false));
 }
 
-#[derive(Fail, Debug)]
+#[derive(ThisError, Debug)]
 pub enum Error {
-    #[fail(display = "{}", _0)]
-    Io(#[cause] std::io::Error),
+    #[error("{0}")]
+    Io(#[from] ::std::io::Error),
```
Comment (Flakebi) [inline src/lib.rs:45]: I think you don’t need the leading `::` with the 2018 edition.

File: src/lib.rs [hunk 3]
```diff
@@ -165,7 +161,7 @@ fn read_u24(inp: &[u8]) -> u32 {
 /// if data would be larger.
 ///
 /// If the incoming data are compressed using level 2, an error is returned.
-pub fn decompress(r: &mut Read, max_size: u32) -> Result<Vec<u8>> {
+pub fn decompress(r: &mut impl Read, max_size: u32) -> Result<Vec<u8>> {
```
Comment (Flakebi) [inline src/lib.rs:164]: Is there a reason to add the `impl`?
I think this makes the function generic, which means bigger binary sizes and longer compile times.
As long as it doesn’t have a measurable impact on performance, I would like to leave it dynamic.

**Review by Flakebi — CHANGES_REQUESTED** [review]
Thanks, looks good in general. I have a few nitpicks inline.

For future prs, it would be nice to split actual changes from formatting changes, either in different prs or different commits. That makes reviewing a lot easier.

**Comment by ariesdevil** [comment]
I'll make a new PR and separate format codes as a new commit

**Pull request closed without merging**